### DISCUSSION TITLE
native pdf visualizer

### DISCUSSION THREAD

**Discussion by Lozamded** [godotengine/godot-proposals#7032] [Engine Core]
### Describe the project you are working on

It would be great to have a node that allows you to view pdf files,

### Describe the problem or limitation you are having in your project

in my case, I work in education developing math applications and games  and sometimes I have to show parts of books,  it would be very useful too to show manuals in remakes of old games

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

A new type of node that you send a file location or a URL , and an index (for the page) and display the content of a pdf file like the containers nodes.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

you create a new node in the control type nodes and someone called pdf container works with the location of the file and the index, you in code can make functions to change the display page index.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

probably if is not useful to many developers can be packaged as a plugin

### Is there a reason why this should be core and not an add-on in the asset library?

because rendering pdf files I think is a relevant feature for the developers in games and and other kind of applications

**Comment by theraot** [comment]
Most games don't need this.

Since Godot does not have this feature. Hopefully you can find launching an external viewer to be a good alternative (with [`OS.execute`](https://docs.godotengine.org/en/stable/classes/class_os.html#class-os-method-execute) or [`OS.create_process`](https://docs.godotengine.org/en/stable/classes/class_os.html#class-os-method-create-process)).

**Reply by Calinou** [reply]
Any web browser is a PDF viewer nowadays. Use `OS.shell_open("path/to/file.pdf")`) and it'll open in the default program associated with PDFs, which is Edge on Windows (which has PDF support).

For paths within `res://` or `user://`, use `OS.shell_open(ProjectSettings.globalize_path("res://path/to/file.pdf"))` for instance.

**Reply by Lozamded** [reply]
I actually do it in that way but now the client allows the ability to add comments to PDFs in the app, I know it's a very specific case, but I don't lose anything by asking.

**Comment by anahuj** [comment]
I need this feature too, but as addon. Pages should be rendered in the app, not in external viewer or in browser.

Would it be trivial to make an addon (gdextension) which includes libpoppler or the like? I was also thinking to use Cygwin+pdftoppm, because I did read that Cygwin can be called with OS.execute() -- check.

My application is a scrapbook... I once did this in GNU/Linux by pre-converting all pdfs to images, which required a lot of disk space.

**Reply by AnasH820** [reply]
man from my experience pdftocairo and pdftoppm are f**king super slow! they take like 1 minute in loading 50 page! and eats memory literally!
